### PR TITLE
#166240967 Add search functionality to the project

### PR DESCRIPTION
**Pull request by eliemugenzi**
### Description of a task to be completed?
A user will be able to search an article by:
- Author
- Title
- Tags
- Keywords

### How should it be manually tested?
- Clone this repo and `cd` into it.
- Run `yarn` or `yarn install` to install and refresh packages
- Run `yarn dev` to run the project
- Click the search icon which redirects you to the search page
- Type in the search box and choose a filter.
- Click the search submit button, then you see the search results.

### Pivotal tracker story
- [#166240967](https://www.pivotaltracker.com/story/show/166240967)

### Screenshot
search

### CODE REVIEW

File: __tests__/components/searchResults.test.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      search: {
+        results: [],
+      },
+    });
+  });
+  test('Test pending branch', () => {
+    results = shallow(<Results {...props2} />);
+    expect(results).toBeDefined();
+  });
+  test('If an author has no articles yet!', () => {
```
Comment (bdushimi) [inline __tests__/components/searchResults.test.js]: @eliemugenzi test case description is not showing the condition under test.

File: __tests__/components/searchResults.test.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+describe('Search results component tests', () => {
+  beforeAll(() => {
+    results = shallow(<Results {...props} />);
+    instance = results.instance();
+  });
+  test('Should render perfectly...', () => {
+    expect(results).toBeDefined();
+  });
+  test('Should receive new props...', () => {
```
Comment (bdushimi) [inline __tests__/components/searchResults.test.js:47]: @eliemugenzi this test does not have assertions!

File: __tests__/components/signup.test.js [hunk 3]
```diff
@@ -101,4 +113,47 @@ describe('Signup Components tests...', () => {
...
+      auth: {
+        signupSuccess: false,
+      },
+    };
+    const component = shallow(<Signup {...propsx} />);
+    const instance = component.instance();
+    it('Should render', () => {
+      expect(component).toBeDefined();
+    });
+    test('It should throw errors', () => {
```
Comment (bdushimi) [inline __tests__/components/signup.test.js]: @eliemugenzi when would it throw the error?

File: __tests__/components/signup.test.js [hunk 4]
```diff
@@ -101,4 +113,47 @@ describe('Signup Components tests...', () => {
...
+        password: '123',
+        confirmPassword: 'whshshs',
+      });
+      instance.forceUpdate();
+      component.update();
+      const event = {
+        preventDefault: jest.fn(),
+      };
+      instance.handleSubmit(event);
+    });
```
Comment (bdushimi) [inline __tests__/components/signup.test.js:152]: @eliemugenzi  add an assertion to this test.

File: __tests__/forms/search.test.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    searchInput.simulate('change', {
+      target: {
+        value: 'deschant',
+      },
+    });
+    const event = {
+      target: {
+        value: 'deschant',
+      },
+    };
```
Comment (bdushimi) [inline __tests__/forms/search.test.js:46]: @eliemugenzi add an assertion to this test.

File: __tests__/forms/search.test.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    const select = search.find('select');
+    select.simulate('change', {
+      target: {
+        value: 'keywords',
+      },
+    });
+    search.instance().onChoose({
+      target: 'keywords',
+    });
+  });
```
Comment (bdushimi) [inline __tests__/forms/search.test.js:59]: @eliemugenzi add an assertion to this test.

File: __tests__/forms/search.test.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    search.instance().forceUpdate();
+    search.update();
+    const event = {
+      preventDefault: jest.fn(),
+    };
+    search.instance().setState({
+      keyword: 'author',
+    });
+    search.instance().onSearch(event);
+  });
```
Comment (bdushimi) [inline __tests__/forms/search.test.js:85]: @eliemugenzi add an assertion to this test

**Review by bdushimi — CHANGES_REQUESTED** [review]

File: __tests__/components/searchResults.test.js (hunk 1 above)
Comment (eliemugenzi) [reply __tests__/components/searchResults.test.js]: The condition is based on the props defined.

**Review by EmyRukundo — APPROVED** [review]

**Comment by deschantkn** [comment]
Hello @eliemugenzi great work implementing the search without having to hit enter to submit. But I'm wondering why there is a 3 character minimum limit to search for a key word, also sometimes there is a message which says `you have made a bad request`, it seems to be random. How is it different from the one that says `You should have provided at least 3 characters long for keywords`?.

File: __tests__/redux/actions/article/searchAction.test.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
+import configureStore from 'redux-mock-store';
+import thunk from 'redux-thunk';
+import moxios from 'moxios';
+import searchArticles from '../../../../src/redux/actions/article/search.action';
+import {
```
[the remark this answers is not in the document]
Comment (eliemugenzi) [reply __tests__/redux/actions/article/searchAction.test.js]: > Hello @eliemugenzi great work implementing the search without having to hit enter to submit. But I'm wondering why there is a 3 character minimum limit to search for a key word, also sometimes there is a message which says `you have made a bad request`, how is it different from the one that says `You should have provided at least 3 characters long for keywords`.

@deschantkn , when is the `You have made a bad request`  being thrown? It should happen if the user makes a remote request with a wrong query parameter. And I think the request I made in my action is valid. Kindly come and show me that case. Thanks for the feedback.

**Comment by eliemugenzi** [comment]
> Hello @eliemugenzi great work implementing the search without having to hit enter to submit. But I'm wondering why there is a 3 character minimum limit to search for a key word, also sometimes there is a message which says `you have made a bad request`, it seems to be random. How is it different from the one that says `You should have provided at least 3 characters long for keywords`?.

**Comment by eliemugenzi** [comment]
@deschantkn , when is the `You have made a bad request`  being thrown? It should happen if the user makes a remote request with a wrong query parameter. And I think the request I made in my action is valid. Kindly come and show me that case. Thanks for the feedback. The minimum of 3 characters is the backend validation.

**Review by Diama1 — APPROVED** [review]

**Comment by UhiriweAudace** [comment]
Kindly @eliemugenzi  resolve those Hound CI issues. Thanks!

**Review by rukundoeric — APPROVED** [review]

**Pull request merged**